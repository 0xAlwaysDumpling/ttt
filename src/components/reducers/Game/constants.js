export const boardSize = 3;
export const initialBoardState = new Array(boardSize).fill(null).map(row => new Array(boardSize).fill(null));

export const initialScoreboard = {
  p1w: 0,
  p2w: 0,
  p1l: 0,
  p2l: 0,
  draws: 0
}

export const initalGameHistoryMapState = (new Array(boardSize).fill(0)).reduce((accum, _, index) => { accum[index] = 0; return accum; }, {});
export const initialGameHistory = {
  hasEnded: false,
  playerturn: 0,
  winner: -1,
  row: { 0: { ...initalGameHistoryMapState }, 1: { ...initalGameHistoryMapState } },
  col: { 0: { ...initalGameHistoryMapState }, 1: { ...initalGameHistoryMapState } },
  diag: { 0: 0, 1: 0 },
  antidiag: { 0: 0, 1: 0 },
}

export const initialState = {
  boardState: initialBoardState,
  gameHistory: initialGameHistory,
  scoreboard: initialScoreboard,
};

export const resetGameState = {
  boardState: initialBoardState,
  gameHistory: initialGameHistory
};
