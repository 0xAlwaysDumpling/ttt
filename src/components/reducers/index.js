import { combineReducers } from 'redux';
import GameState from './Game';


export default combineReducers({
  GameState
});

