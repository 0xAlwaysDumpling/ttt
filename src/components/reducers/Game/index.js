
import { loadState } from '../../localstorage';
import { initialState, resetGameState } from './constants';
import { postMove } from '../../api';

/* 
  To mock API call I needed to do actions MOVE_REQUEST, MOVE_SUCCESS, MOVE_FAILED 
  In the interest of scope and time I just made a function call postMove

*/
const GameState = (state = initialState, action) => {
  switch (action.type) {
    case 'LOAD':
      const appState = loadState();
      return appState ? appState : initialState;
    case 'MOVE':
      if (state.gameHistory.hasEnded) return state;
      else return postMove(state, action)
    case 'RESET':
      return {
        ...state,
        ...resetGameState
      };
    default:
      return state;
  }
}

export default GameState;