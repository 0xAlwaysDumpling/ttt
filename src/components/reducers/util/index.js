import React from 'react'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import { render } from '@testing-library/react';
import rootReducer from '../index';
import { initialState as GameInitialState } from '../Game/constants'

const initialState = {
  GameState: GameInitialState
}

export const renderWithRedux = (
  component,
) => {
  const store = createStore(rootReducer, initialState)
  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  }
}
