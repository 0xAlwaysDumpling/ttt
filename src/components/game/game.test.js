import React from 'react'
import { cleanup } from '@testing-library/react'
import Game from './index';
import { renderWithRedux } from '../reducers/util';

afterEach(cleanup)

it('should render with out crashing', () => {
  renderWithRedux(<Game />)
});

it('should take a snapshot of Game', () => {
  const { asFragment } = renderWithRedux(<Game />)

  expect(asFragment(<Game />)).toMatchSnapshot()
});
