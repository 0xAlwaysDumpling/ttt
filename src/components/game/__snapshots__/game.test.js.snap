// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot of Game 1`] = `
<DocumentFragment>
  <div
    class="sc-AxhUy eErOKf"
  >
    <div
      class="sc-AxirZ hJmkar"
    >
      <div
        class="sc-AxiKw CYfUp"
      >
        <div
          class="sc-AxhCb exkQbL"
        >
          <button
            class="sc-AxjAm dRYkuE"
          />
        </div>
        <div
          class="sc-AxhCb exkQbL"
        >
          <button
            class="sc-AxjAm dRYkuE"
          />
        </div>
        <div
          class="sc-AxhCb exkQbL"
        >
          <button
            class="sc-AxjAm dRYkuE"
          />
        </div>
      </div>
      <div
        class="sc-AxiKw CYfUp"
      >
        <div
          class="sc-AxhCb exkQbL"
        >
          <button
            class="sc-AxjAm dRYkuE"
          />
        </div>
        <div
          class="sc-AxhCb exkQbL"
        >
          <button
            class="sc-AxjAm dRYkuE"
          />
        </div>
        <div
          class="sc-AxhCb exkQbL"
        >
          <button
            class="sc-AxjAm dRYkuE"
          />
        </div>
      </div>
      <div
        class="sc-AxiKw CYfUp"
      >
        <div
          class="sc-AxhCb exkQbL"
        >
          <button
            class="sc-AxjAm dRYkuE"
          />
        </div>
        <div
          class="sc-AxhCb exkQbL"
        >
          <button
            class="sc-AxjAm dRYkuE"
          />
        </div>
        <div
          class="sc-AxhCb exkQbL"
        >
          <button
            class="sc-AxjAm dRYkuE"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
